### ISSUE TITLE
Training clarification

### ISSUE THREAD

**Issue by sampepose** [lmb-freiburg/demon#10]
Hello!

I want to finetune your network on my dataset. I have a couple questions about how you initially trained the network.

First you trained the four encoder-decoders in the bootstrap and iterative net sequentially for 250k iterations. This is straightforward.

Second you trained the iterative net for 1600k iterations. You appended the predictions of the iterative net to the inputs to the iterative net to emulate having 4x iterative nets while only having one net in memory.

- Is the loss only defined / backpropagated with the *final* predictions (after the 4x iterations)? Or is the loss backpropagated after *each* iteration? 
- If the loss is only backpropagated after all four iterations of the iterative net, is a single minibatch iteration (in the sense that it was trained for 1600k iterations) defined as one forward-pass of the bootstrap and iterative nets or 4x forward passes of the bootstrap and iterative nets (a single backwards pass)?
- I'm unclear about what this means: "We also train the first iteration on its own, but then train all iterations jointly which avoids intermediate storage." Is this referring to when you trained the iterative net as part of the first 250k iterations?

Finally, do you see any need to finetune the refinement net or is it sufficient to finetune the iterative net with simulated iterations?

Thank you!

Sam Pepose

**Comment by benjaminum** [comment]
Hi,

> Is the loss only defined / backpropagated with the final predictions (after the 4x iterations)? Or is the loss backpropagated after each iteration?

The loss/gradient is computed for the whole batch of predictions, and the batch contains predictions from the iterations 1-4 of the iterative net.

> If the loss is only backpropagated after all four iterations of the iterative net, is a single minibatch iteration (in the sense that it was trained for 1600k iterations) defined as one forward-pass of the bootstrap and iterative nets or 4x forward passes of the bootstrap and iterative nets (a single backwards pass)?

1600k iterations means forward/backward passes during training.

> I'm unclear about what this means: "We also train the first iteration on its own, but then train all iterations jointly which avoids intermediate storage." Is this referring to when you trained the iterative net as part of the first 250k iterations?

_all iterations jointly_ refers to the input batch to the iterative network which contains predictions from the bootstrap net and the predictions of the iterative net of iterations 1-3.

> Finally, do you see any need to finetune the refinement net or is it sufficient to finetune the iterative net with simulated iterations?

It depends on your data. I would expect reasonable results without finetuning the refinement net.

**Comment by sampepose** [comment]
Thanks, Benjamin! That makes sense.
